[KEYCLOAK-11621] Allow user creation via group MANAGE_MEMBERS and MANAGE_MEMBERSHIP permissions (Admin API)

clementcur opened the pull request:
Jira Issues:
- [KEYCLOAK-11621](https://issues.redhat.com/browse/KEYCLOAK-11621)

Problem:
Using fine-grained admin permissions on groups, it is not permitted to create new users
within a group.

Cause:
The POST /{realm}/users API does not check permission for each group part of the new
user representation

Solution:
- Change access logic for POST /{realm}/users to require MANAGE_MEMBERS and
MANAGE_MEMBERSHIP permissions on each of the incoming groups

Tests:
Manual API testing performed:
  1. admin user from master realm:
    - POST /{realm}/users without groups                  => HTTP 201 user created
    - POST /{realm}/users with groups                     => HTTP 201 user created
  2. user with MANAGE_MEMBERS & MANAGE_MEMBERSHIP permissions on group1
    - POST /{realm}/users without groups                  => HTTP 403 user NOT created
    - POST /{realm}/users with group1                     => HTTP 201 user created
    - POST /{realm}/users with group1 & group2            => HTTP 403 user NOT created
    - POST /{realm}/users with group1 & wrong group path    => HTTP 400 user NOT created
  3. user with MANAGE_MEMBERS permission on group1
    - POST /{realm}/users without groups                  => HTTP 403 user NOT created
    - POST /{realm}/users with group1                     => HTTP 403 user NOT created
    - POST /{realm}/users with group1 & group2            => HTTP 403 user NOT created
    - POST /{realm}/users with group1 & wrong group path    => HTTP 400 user NOT created

Commit:
[KEYCLOAK-11621] Allow user creation via group permissions (Admin API)

Problem:
Using fine-grained admin permissions on groups, it is not permitted to create new users
within a group.

Cause:
The POST /{realm}/users API does not check permission for each group part of the new
user representation

Solution:
- Change access logic for POST /{realm}/users to require MANAGE_MEMBERS and
MANAGE_MEMBERSHIP permissions on each of the incoming groups

Tests:
Manual API testing performed:
  1. admin user from master realm:
    - POST /{realm}/users without groups                  => HTTP 201 user created
    - POST /{realm}/users with groups                     => HTTP 201 user created
  2. user with MANAGE_MEMBERS & MANAGE_MEMBERSHIP permissions on group1
    - POST /{realm}/users without groups                  => HTTP 403 user NOT created
    - POST /{realm}/users with group1                     => HTTP 201 user created
    - POST /{realm}/users with group1 & group2            => HTTP 403 user NOT created
    - POST /{realm}/users with group1 & wrong group path  => HTTP 400 user NOT created
  3. user with MANAGE_MEMBERS permission on group1
    - POST /{realm}/users without groups                  => HTTP 403 user NOT created
    - POST /{realm}/users with group1                     => HTTP 403 user NOT created
    - POST /{realm}/users with group1 & group2            => HTTP 403 user NOT created
    - POST /{realm}/users with group1 & wrong group path  => HTTP 400 user NOT created

stianst commented on services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java:
```diff
@@ -105,7 +100,21 @@ public UsersResource(RealmModel realm, AdminPermissionEvaluator auth, AdminEvent
+                GroupModel group = realm.getGroupById(groupId);
+                if (group != null) {
+                    auth.groups().requireManageMembers(group);
+                    auth.groups().requireManageMembership(group);
```
This is strange, what's the difference between requireManageMembers and requireManageMembership?

clementcur replied:
My understanding is the following:
- MANAGE_MEMBERS_SCOPE: allowing an admin to manage users, i.e. create, update and delete users
- MANAGE_MEMBERSHIP_SCOPE: allowing an admin to manage groups membership for a specific user

So, when creating a new user part of one or more groups, we need both permissions IMO. Let me know your thoughts.

pedroigor replied:
Yeah, it makes sense. 

The `requireManageMembers` works as an alternative to `users().requireManage()` so that admins can manage users from a specific group. You are introducing (and probably fixing if we don't have an issue yet) the possibility to also create users when the admin has `MANAGE_MEMBERS_SCOPE` for a group. By doing this, you should also check the permissions when updating and deleting users. As well as provide tests to cover the behavior.

As a suggestion, I would not introduce this behavior but just create the user and assign it to a group. You could keep the `MANAGE_MEMBERSHIP_SCOPE` though.

stianst commented:
The approach of allowing adding a user with membership to groups (where there is permission to manage membership) is okay when looking at it purely from a REST API perspective. However, we also need to consider how this would fit into the Keycloak admin console.

Do you have any thoughts with regards to the Keycloak admin console?

clementcur commented:
Hi @stianst, I made an initial proposal on the Jira ticket #KEYCLOAK-11621. I don't mind contributing for the UI part too.

1. **Tabular view**: 
  _Option 1:_
  -> in Users table view, add a dropdown action box in top right corner (see attachment [users_table.PNG](https://issues.redhat.com/secure/attachment/12463591/users_table.PNG))
    - only visible when both "MANAGE_MEMBERS_SCOPE" and "MANAGE_MEMBERSHIP_SCOPE" permissions are set on at least one group for the connected user, and no admin rights (otherwise Add User classic button showing)
    - with the following values:
      - Create user in group... (unselectable)
      - Create user in My First Group
      - Create user in Another Group I am Allowed to Manage
    - will pre-fill the Add User form field "Groups" with the one selected
    - the API will only allow user creation if the Groups field has one or more authorized groups

    _Option 2:_
    -> in Users table view, change visibility of the existing "Add User" dropdown action box in top right corner (no wireframe needed, since no change from current Users table view)
      - only visible when both "MANAGE_MEMBERS_SCOPE" and "MANAGE_MEMBERSHIP_SCOPE" permissions are set on at least one group for the connected user, or admin rights
      - no pre-fill on the Add User form field "Groups"
      - the API will only allow user creation if the Groups field has one or more authorized groups

2. **Detailed view (create and update user form)**:
  -> add a multi-valued selectable field for assigning groups (see attachment [users_add.PNG](https://issues.redhat.com/secure/attachment/12463590/users_add.PNG))
    - no tree browsing, only type-and-match
    - pre-filled with the group selected in the dropdown box previsouly (for proposal option 1)

slaskawi commented:
@davidffrench @pb82 This is something we could probably use in Keycloak Operator.

stianst commented:
Your proposal works for me and I'm happy with getting this merged first. However, I'd like a new JIRA that covers just this PR (ability to create users within a group via admin api). Can you create a JIRA and change the commit message?

stianst commented:
I'm wondering about the UI part if it should just wait until the new admin console. This year we are planning to rewrite the admin console using React and PatternFly 4, so would be a bit of waste of time doing it for the old one unless that's something you want?

clementcur commented:
@stianst yeah, it makes sense to link this PR to a JIRA covering only the API part. I guess the existing [KEYCLOAK-12416](https://issues.redhat.com/browse/KEYCLOAK-12416) issue is a better fit. I'll rename the PR then.

About the UI part, that's something we'll need pretty soon. So I'd vote for getting the UI in sync with the API asap (I can work on a PR in the coming weeks), and keep that feature in mind when building the new UI... what do you think?

clementcur commented on services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java:
```diff
@@ -105,7 +100,21 @@ public UsersResource(RealmModel realm, AdminPermissionEvaluator auth, AdminEvent
+                    auth.groups().requireManageMembership(group);
+                } else {
+                    // unrecognized group from id
+                    throw new ForbiddenException();
```
yeah, makes sense - will update the PR then

stianst commented:
@clementcur sounds good - as we are now past code freeze for 9, I'm going to put a hold on this PR for a bit and schedule it for 10.

clementcur commented:
ping @stianst, any news?

clementcur commented:
I've updated my PR since part of it has actually been merged via PR #6886. So now, this PR only deals with allowing user creation via fine grain admin permissions.

@stianst, let me know if this can make it into the next major release.

clementcur commented:
> I'm wondering about the UI part if it should just wait until the new admin console. This year we are planning to rewrite the admin console using React and PatternFly 4, so would be a bit of waste of time doing it for the old one unless that's something you want?

@stianst I've posted a new PR for the UI part, see https://github.com/keycloak/keycloak/pull/7035. Note that it'll require another PR to allow finegrain group admins to see the "Add user" button in the list view and the "Save" button in the creation details view.

stianst commented:
Sorry for the time it's taken to respond to this issue.

Can you send an email to the dev mailing list to discuss this (as well as https://github.com/keycloak/keycloak/pull/7035)? So we can get input from others.

clementcur commented:
done @stianst: https://groups.google.com/d/topic/keycloak-dev/mOxO_ET4g2w/discussion

pedroigor commented on services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java:
```diff
@@ -107,7 +109,20 @@ public UsersResource(RealmModel realm, AdminPermissionEvaluator auth, AdminEvent
     @POST
     @Consumes(MediaType.APPLICATION_JSON)
     public Response createUser(final UserRepresentation rep) {
-        auth.users().requireManage();
```
I did not catch this before, but I think we should first try to check `requireManage` and only in case we get an exception (lacking permission) we fallback to your changes that rely on the fine-grained permission.

The reason behind this comment is that we avoid people not using fine-grained permissions to not be impacted by the additional checks you are adding.

pedroigor reviewed (changes requested):
@clementcur LGTM as an initial step to support creating users wh

pedroigor commented:
@clementcur Despite my last comment and the fact that you don't have yet provided a test, it LGTM.

For testing this, please take a look at `org.keycloak.testsuite.admin.FineGrainAdminUnitTest`.

clementcur commented:
ok, thx for your feedback @pedroigor. I'll go ahead and add some tests then.

Commit:
Finegrain admin as fallback and added some tests

clementcur commented:
Hi @pedroigor, I've changed the logic as you suggested : fallingback to finegrain admin permissions only if requireManage fails. Also, I've added some integration tests. Let me know if I'm missing anything else.
